numba/tests/test_parfors.py

```diff
@@ -2039,6 +2039,20 @@ def test_impl(n):
             return B
         self.prange_tester(test_impl, 1.0)
 
+    @skip_unsupported
+    def test_mutable_list_param(self):
+        def test_impl(n):
+            def list_check(X):
+                a = X[-1] + 1
+                if a not in X:
+                    X.append(a)
+                return a
+            for i in range(n):
+                X = [100]
+                a = list_check(X)
+            return a
+        self.prange_tester(test_impl, 10)
```

Comment:
I think this compiles with or without the fix and it seems like `X` isn't hoisted irrespective of the fix being present, I guess because `list_check` is an inlinable function it's inlined at the call site by the closure inlining pass so there's something that prevents hoisting now present? The function also has a race condition, but I don't think that matters. The parfors diagnostic metadata may help in writing a test:
```python
from numba import prange, njit

@njit
def list_check(X):
        a = X[-1] + 1
        if a not in X:
            X.append(a)
        return a

@njit(parallel=True)
def test_impl(n):
    for i in prange(n):
        X = [100]
        a = list_check(X)
    return a

test_impl(10)

diagnostics = test_impl.get_metadata()[test_impl.signatures[0]]['parfor_diagnostics']
diagnostics.dump(4)

for loop, hoist_info in diagnostics.hoist_info.items():
    for h, stmts in hoist_info.items():
        print("Info %s" % h)
        for stmt in stmts:
            print("loop %s. %s" % (loop, stmt))
        print()
```

Reply:
I changed list_check slightly so that it would definitely fail if X = [100] were hoisted.  I also confirmed manually with parfor diagnostics that it isn't hoisted here.  I switched this test to manual style instead of prange_tester.